File: ActionUI/Common/WindowModal.swift
// Common/WindowModal.swift

import SwiftUI

// MARK: - Public types (used in ActionUIModel public API)

/// Presentation style for window-level modal.
public enum ModalStyle: Sendable {
    case sheet
    case fullScreenCover
}

/// A button descriptor for window-level alerts and confirmation dialogs.
public struct DialogButton: Sendable {
    public let title: String
    public let role: ButtonRole?    // nil = default, .destructive, .cancel
    public let actionID: String?    // action fired on tap; nil = dismiss only

    public init(title: String, role: ButtonRole? = nil, actionID: String? = nil) {
        self.title = title
        self.role = role
        self.actionID = actionID
    }
}

// MARK: - Internal model types

/// Differentiates alert from confirmationDialog inside WindowDialog.
enum DialogStyle {
    case alert
    case confirmationDialog
}

/// Holds a loaded element and its presentation style for window-level sheet / fullScreenCover.
/// Created by ActionUIModel.presentModal; dismissed by ActionUIModel.dismissModal.
@MainActor
struct WindowModal: Identifiable {
    let id = UUID()
    let element: ActionUIElement
    let style: ModalStyle
    let onDismissActionID: String?
    /// IDs of all ViewModels loaded for this modal — removed from WindowModel on dismiss.
    let loadedViewIDs: Set<Int>
}

/// Holds pure data for a window-level alert or confirmationDialog.
/// No ViewModels are allocated — content is title, message, and buttons only.
struct WindowDialog: Identifiable {
    let id = UUID()
    let style: DialogStyle
    let title: String
    let message: String?
    let buttons: [DialogButton]
}

File: ActionUIAndroid/library/src/main/java/com/abracode/actionui/Common/WindowModal.kt
package com.abracode.actionui.Common

/**
 * Window-level modal support - the Android port of the *modal* half of Apple's
 * `WindowModal.swift`: the `sheet` / `fullScreenCover` presentation driven by
 * `ActionUIModel.presentModal` / `dismissModal` and rendered by
 * [com.abracode.actionui.Helpers.WindowModalHost].
 *
 * Unlike a [WindowDialog] (pure title / message / buttons), a modal loads a JSON
 * sub-document, registers its [ViewModel]s into the window pool so the value / state
 * API reaches the modal's controls, and removes exactly those on dismiss. See
 * `Private/Android_Porting_Notes.md`.
 */

/** Presentation style for a window-level modal. Mirror of Swift's `ModalStyle`. */
enum class ModalStyle { SHEET, FULL_SCREEN_COVER }

/**
 * An active window-level modal: the loaded [element], its presentation [style], an
 * optional [onDismissActionID] fired once when it closes, and [loadedViewIDs] - the
 * element ids whose [ViewModel]s this modal added to the window pool (removed on
 * dismiss). Held by [WindowModel.windowModal]; created by
 * `ActionUIModel.presentModal`, cleared by `dismissModal`. Mirror of Swift's
 * `WindowModal`.
 */
data class WindowModal(
    val element: ActionUIElement,
    val style: ModalStyle,
    val onDismissActionID: String? = null,
    val loadedViewIDs: Set<Int>,
)
